Summing compaction fixes

vavrusa opened the pull request:
This fixes several problems in test related to compaction:

* Keys with zero values are compacted in `sumMap`
* Fixed SummingMergeTree in case an explicit list of columns to sum is given in configuration
* Summation of non-UInt64 columns is fixed, the aggregation now uses `sumWithOverflow(x)` which is a variant of `sum(x)` function that uses same data type for return value as the input type

vavrusa committed:
AggregateFunctionSumMap: compact keys with zero value

This allows `sumMap(k, v)` to compact all keys
for which the values are all zero.

vavrusa committed:
SummingSortedBlockInputStream: fix explicitly configured columns to sum

This fixes a variant of SummingMergeTree() in which
the columns to sum are explicitly configured.
Previously columns not in that list were ignored,
instead of writing last value.

This also fixes summation of invalid maps with
with only one key column and no value columns.

Modified test to work around compaction limitation
in which a zero-value column isn’t compacted
immediately if the inputs are non-zero but the
output is zero (+1 -1).

vavrusa committed:
AggregateFunctionSum: implemented sumWithOverflow(x)

By default sum(x) promotes the result type to
largest possible integral type to avoid
arithmetic overflow when summing values from
smaller data types.

This is generally desirable behaviour, but it doesn’t
work with summing merge tree, as the result is
expected to be of same type as the input.

alexey-milovidov commented on dbms/src/AggregateFunctions/AggregateFunctionSum.cpp:
```diff
@@ -26,6 +39,7 @@ AggregateFunctionPtr createAggregateFunctionSum(const std::string & name, const
 void registerAggregateFunctionSum(AggregateFunctionFactory & factory)
 {
     factory.registerFunction("sum", createAggregateFunctionSum, AggregateFunctionFactory::CaseInsensitive);
+    factory.registerFunction("sumWithOverflow", createAggregateFunctionSumWithOverflow, AggregateFunctionFactory::CaseInsensitive);
```
Don't need to be case insensisive.

alexey-milovidov commented on dbms/src/AggregateFunctions/AggregateFunctionSum.cpp:
```diff
@@ -13,6 +13,19 @@ AggregateFunctionPtr createAggregateFunctionSum(const std::string & name, const
+    if (argument_types.size() != 1)
+        throw Exception("Incorrect number of arguments for aggregate function " + name, ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
     AggregateFunctionPtr res(createWithNumericType<AggregateFunctionSum>(*argument_types[0]));
```
?

vavrusa commented on dbms/src/AggregateFunctions/AggregateFunctionSum.cpp:
```diff
@@ -13,6 +13,19 @@ AggregateFunctionPtr createAggregateFunctionSum(const std::string & name, const
     if (argument_types.size() != 1)
         throw Exception("Incorrect number of arguments for aggregate function " + name, ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
 
+    AggregateFunctionPtr res(createWithNumericTypeNearest<AggregateFunctionSum>(*argument_types[0]));
```
This just instantiates a different variant of the template with `AggregateFunctionSum<T, NearestFieldType<T>::Type>`.

ludv1x commented:
autotests

vavrusa commented:
@ludv1x can you repost which tests have failed?

vavrusa committed:
tests: added tests for sumWithOverflow(x) preserving input type

ludv1x commented:
@vavrusa 
stateless.00155_merges
```
Received exception from server:
Code: 9. DB::Exception: Received from localhost:9000, ::1. DB::Exception: Sizes of columns doesn't match: d: 2, s: 0. 
Received exception from server:
Code: 9. DB::Exception: Received from localhost:9000, ::1. DB::Exception: Sizes of columns doesn't match: d: 2, s: 0. 
Received exception from server:
Code: 9. DB::Exception: Received from localhost:9000, ::1. DB::Exception: Sizes of columns doesn't match: d: 2, s: 0. 
```

This test is in public part of the test suite (`queries/0_statetles/00155...`).

vavrusa commented:
@ludv1x is there a script to run these manually or as part of the build?

vavrusa committed:
SummingSortedBlockInputStream: fix summing of multiple blocks

Previously the destination columns were only computed for the first
block, so that subsequently written blocks failed to write
aggregation results to corrent columns.

vavrusa commented:
@ludv1x thanks Vitaliy, that was super helpful, I fixed the mistake.

ludv1x commented:
> is there a script to run these manually or as part of the build?

`./release` the script only builds a deb package, so you should run `./clickhouse-test` manually.

vavrusa commented:
Brilliant, this is great, thanks @ludv1x

```
$ /usr/local/bin/python2 ../dbms/tests/clickhouse-test -b "$(pwd)/clickhouse-server" -c "$(pwd)/clickhouse-client" --no-zookeeper --no-shard -q $(pwd)/../dbms/tests/queries "(sum|merges)"

Running stateless tests.

00507_sumwithoverflow:                                                 [ OK ]
00502_sum_map:                                                         [ OK ]
00327_summing_composite_nested:                                        [ OK ]
00322_disable_checksumming:                                            [ OK ]
00155_merges:                                                          [ OK ]
00148_summing_merge_tree_nested_map_multiple_values:                   [ OK ]
00146_summing_merge_tree_nested_map:                                   [ OK ]
00084_summing_merge_tree:                                              [ OK ]
00043_summing_empty_part:                                              [ OK ]

All tests passed.
```

vavrusa commented:
@alexey-milovidov can you take a look? This should fix autotests again.
